### ISSUE TITLE
conda-build not seeming to pick up git env variables

### ISSUE THREAD

**Issue by jreback** [conda/conda-build#764]
conda-build 1.19

I recall this working previously (3 mo ago)

``conda build conda.recipe --no-test --numpy 1.10 --python 2.7``

```
package:
    name: pandas
    version: {{ environ.get('GIT_DESCRIBE_TAG', '') }}

build:
    number: {{ environ.get('GIT_DESCRIBE_NUMBER', 0) }}
    string: np{{ environ.get('NPY_VER').replace('.', '') }}py{{ environ.get('PY_VER').replace('.', '') }}_{{ environ.get('GIT_BUILD_STR', '') }}

source:
    git_url: ../

requirements:
  build:
    - python
    - cython
    - numpy x.x
    - setuptools
    - pytz
    - python-dateutil

  run:
    - python
    - numpy x.x
    - python-dateutil
    - pytz

test:
  imports:
    - pandas

about:
  home: http://pandas.pydata.org
  license: BSD
```

**Comment by msarahan** [comment]
Ping @stuarteberg - I think you know the most about this stuff.  It seems that none of these GIT_DESCRIBE_ variables are getting defined.  What are the requirements for those to be defined?

**Comment by stuarteberg** [comment]
I don't know exactly why conda-build 1.19 doesn't work for that recipe, but I did notice that my recent changes in #758 will also have trouble with a recipe like this.  Hence, I just submitted #765.

>What are the requirements for those to be defined?

The only requirement is that you define `git_url`.  As of 7689ae1e8a8965cec405f2a4db3f1f94c5d02ce5 (and a little fix in fddfb3209a85be286f013006f8c687e5a165ee72), the `path` field cannot be used if you want to use `GIT_` variables -- you have to use `git_url`.

**Comment by jreback** [comment]
@stuarteberg I just tried with conda-build master (where #765) is merged, does not appear to fix this issue (with my given recipe). Is there something I can tweek in the recipe to make the variables work?

**Comment by jreback** [comment]
```
conda-build               1.19.0+13.g48c3b7a           <pip>
```

```
[jreback-~/pandas] conda build conda.recipe --no-test --python 2.7 --numpy 1.10
Removing old build environment
Removing old work directory
BUILD START: pandas--np110py27_
Fetching package metadata: ................
WARNING: Could not detect installed version of conda-build
Solving package specifications: ................
The following NEW packages will be INSTALLED:

    cython:          0.23.4-py27_1
    mkl:             11.3.1-0     
    numpy:           1.10.4-py27_0
    openssl:         1.0.2d-0     
    pip:             8.0.2-py27_0 
    python:          2.7.11-0     
    python-dateutil: 2.4.2-py27_0 
    pytz:            2015.7-py27_0
    readline:        6.2-2        
    setuptools:      19.6.2-py27_0
    six:             1.10.0-py27_0
    sqlite:          3.9.2-0      
    tk:              8.5.18-0     
    wheel:           0.29.0-py27_0
    zlib:            1.2.8-0      

Linking packages ...
```

**Comment by stuarteberg** [comment]
It works for me, but I think I figured out why it's not working for you.  Your recipe provides `git_url`, but does not provide a `git_tag`.  Since `git_tag` is missing, we assume you meant `master`.  But if you are currently testing via a branch, then the commit of `conda-bld/work` won't match the name `master`.  My changes in #765 require that the checkout git tag matches the `git_tag` from your `meta.yaml` file, or else it refuses to define `GIT_DESCRIBE_TAG`, etc.

So... does it work if you add `git_tag` to your recipe?

**Comment by stuarteberg** [comment]
BTW, you can simplify the jinja templates in your `meta.yaml` file:

```yaml
package:
    name: pandas
    version: {{ GIT_DESCRIBE_TAG.replace('.dev', 'dev') }} # Requires #771

build:
    number: {{ GIT_DESCRIBE_NUMBER|int }}
    string: np{{ CONDA_NPY }}py{{ CONDA_PY }}_{{ GIT_BUILD_STR }}

```

Well, the syntax in the `version` field is using `.replace`, which won't work until #771 is merged.  But the rest of it already works.

**Comment by jreback** [comment]
ok, the changes to ``meta.yaml`` look good (I don't need the ``.dev`` in any event). But still same issue
no mater what I try for ``git_tag``, e.g. nothing, 'master', the actual hash, or even a version number

**Comment by stuarteberg** [comment]
Well I'm stumped then.  For the record, I'm testing with my pandas fork (branch [test-conda](https://github.com/stuarteberg/pandas/tree/test-conda)).  Does the following work for you?

```bash
cd pandas
git remote add stuart https://github.com/stuarteberg/pandas
git fetch stuart
git checkout test-conda
conda build conda.recipe --no-test --python 2.7 --numpy 1.10
```

Here's the output I get.  Note that the jinja template string will be unresolved on the `BUILD_START` line, but later it gets fixed (after the source has been cloned).

```
Removing old build environment
Removing old work directory
BUILD START: pandas--np110py27_0
Fetching package metadata: ........
Solving package specifications: ................
The following NEW packages will be INSTALLED:

    cython:          0.23.4-py27_1
    mkl:             11.3.1-0     
    numpy:           1.10.4-py27_0
    openssl:         1.0.2f-0     
    pip:             8.0.2-py27_0 
    python:          2.7.11-0     
    python-dateutil: 2.4.2-py27_0 
    pytz:            2015.7-py27_0
    readline:        6.2-2        
    setuptools:      19.6.2-py27_0
    six:             1.10.0-py27_0
    sqlite:          3.9.2-0      
    tk:              8.5.18-0     
    wheel:           0.29.0-py27_0
    zlib:            1.2.8-0      

Linking packages ...
[      COMPLETE      ]|###################################################################################################################################################| 100%
Removing old work directory
checkout: u'test-conda'
Cloning into '/miniconda/conda-bld/work'...
done.
Branch test-conda set up to track remote branch test-conda from origin.
Switched to a new branch 'test-conda'
==> git log -n1 <==

commit f3bf0f08db7a2c31b32d734731d1e57b9e3ecdb1
Author: Stuart Berg <bergs@janelia.hhmi.org>
Date:   Tue Feb 9 16:45:55 2016 -0500

    testing conda recipe


==> git describe --tags --dirty <==

v0.17.1-304-gf3bf0f0


==> git status <==

On branch test-conda
Your branch is up-to-date with 'origin/test-conda'.
nothing to commit, working directory clean


Package: pandas-v0.17.1-np110py27_304
source tree in: /miniconda/conda-bld/work
number of files: 0
Fixing permissions
Fixing permissions
BUILD END: pandas-v0.17.1-np110py27_304
# If you want to upload this package to anaconda.org later, type:
#
# $ anaconda upload /miniconda/conda-bld/osx-64/pandas-v0.17.1-np110py27_304.tar.bz2
#
# To have conda build upload to anaconda.org automatically, use
# $ conda config --set anaconda_upload yes
```

**Comment by jreback** [comment]
ahh maybe that's it
I was thinking the build start line should be right

ok let me try again

**Comment by jakirkham** [comment]
Unfortunately, not. It fills these variables the same way that we might do ourselves. Namely calling `git` with some subcommands in the repo. So, it won't be able to do that until there is a cloned repo in a directory where it can run `git`.

**Comment by jreback** [comment]
ok this works great thanks!
``BUILD END: pandas-v0.17.1-np110py27_321_ga66a3e3``

only request would be to potentially show a better BUILD START version (not sure if easy/possible)

I'll leave open but close if you deem this closed.

**Comment by stuarteberg** [comment]
>only request would be to potentially show a better BUILD START version (not sure if easy/possible)

It's not really possible.

>I'll leave open but close if you deem this closed.

Please close.

**Comment by jakirkham** [comment]
Well, couldn't one make the argument that "BUILD START" could be printed after the source code is gotten and placed in the working build directory? Whether that should be the case is a question for another issue. Personally, I don't care so much, but I suppose it is worth considering.

**Comment by jreback** [comment]
unrelated, but getting this is 3.4 (3.5 and 2.7 work fine), any ideas?
```
install_name_tool -add_rpath @loader_path/../../../ /Users/jreback/miniconda/envs/_build/lib/python3.4/site-packages/pandas/parser.so
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/install_name_tool: changing install names or rpaths can't be redone for: /Users/jreback/miniconda/envs/_build/lib/python3.4/site-packages/pandas/parser.so (for architecture x86_64) because larger updated load commands do not fit (the program must be relinked, and you may need to use -headerpad or -headerpad_max_install_names)

Error: install_name_tool failed with exit status 1
[jreback-~/pandas] conda build conda.recipe --no-test --python 3.4 --numpy 1.10
```

**Comment by jakirkham** [comment]
Could you please put that in a different issue?

**Comment by stuarteberg** [comment]
>couldn't one make the argument that "BUILD START" could be printed after the source code is gotten and placed in the working build directory?

No. The point of the `BUILD START` line is that it makes it clear which package conda is currently working on.  If you get an error (say, while parsing meta.yaml or downloading the source), you should be able to see which recipe is having trouble based on the console output.  Moving the `BUILD START` line would cause confusion.

**Comment by jreback** [comment]
sure

**Comment by jakirkham** [comment]
Thanks.

**Comment by jakirkham** [comment]
> No. The point of the BUILD START line is that it makes it clear which package conda is currently working on....Moving the BUILD START line would cause confusion.

Then you just add a new description like ACQUIRING SOURCE START. I really don't see the problem.

> If you get an error (say, while parsing meta.yaml or downloading the source), you should be able to see which recipe is having trouble based on the console output.

Honestly, the yaml parsing has gotten complicating (and potentially confusing) enough that it may deserve a separate section from the build.

**Comment by stuarteberg** [comment]
>unrelated, but getting this is 3.4 (3.5 and 2.7 work fine), any ideas?

It's complaining that the build PREFIX was too short, so it's not able to replace that string in the binary files when conda performs it's "relocation" step (in this case, via `install_name_tool`.  I have no idea why it would only happen to you in 3.4 and not 2.7 or 3.5.  But try adding this to your recipe:

    build:
      detect_binary_files_with_prefix: True

That will cause conda's temporary build environment PREFIX to be something really long instead of merely `envs/_build`.  (In my case, it becomes `envs/_build_placehold_placehold_placehold_placehold_placehold_placeho`.)

**Comment by msarahan** [comment]
Closed by #758, #759, and #923
